### PR TITLE
[REF] product_supplierinfo_for_customer_sale: Return all result in name_search, products found by default code and products found by customer code

### PR DESCRIPTION
**Pull request by ysantiago**

### CODE REVIEW

**Comment by ysantiago** [comment]
@JulioSerna @moylop260 can you please review this change?

File: product_supplierinfo_for_customer_sale/models/product_product.py [hunk 1]
```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
...
+        if not name or not self._context.get('partner_id'):
             return res
         partner_id = self._context['partner_id']
         supplierinfo = self.env['product.supplierinfo'].search(
             [('name', '=', partner_id), '|', ('product_code', operator, name),
              ('product_name', operator, name)])
         product = self.search(
-            [('product_tmpl_id', '=', supplierinfo.product_tmpl_id.id)],
+            [('product_tmpl_id', 'in',
+              supplierinfo.mapped('product_tmpl_id.id'))],
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/models/product_product.py]: `s/supplierinfo.mapped('product_tmpl_id.id')/supplierinfo.mapped('product_tmpl_id').ids/g`

File: product_supplierinfo_for_customer_sale/models/product_product.py [hunk 2]
```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
...
             limit=limit)
-        res = product.name_get()
-        return res
+        # list(set()) to delete repeated records (if exist)
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/models/product_product.py]: Remove this comment (The code is obvious that delete repeated)

File: product_supplierinfo_for_customer_sale/models/product_product.py [hunk 3]
```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
...
+        return list(set(res + product.name_get()))
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/models/product_product.py]: Use `res.extend(product.name_get())` in order to use the same pointer

**Review by moylop260 — CHANGES_REQUESTED** [review]
Please, add a test case where a result returns 2 or more items

**Comment by ysantiago** [comment]
@moylop260 please, can you see the changes?

**Review by moylop260 — APPROVED** [review]

File: product_supplierinfo_for_customer_sale/models/product_product.py [hunk 4]
```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
...
              ('product_name', operator, name)])
         product = self.search(
-            [('product_tmpl_id', '=', supplierinfo.product_tmpl_id.id)],
+            [('product_tmpl_id', 'in',
+              supplierinfo.mapped('product_tmpl_id').ids)],
             limit=limit)
-        res = product.name_get()
-        return res
+        res.extend(product.name_get())
+        return list(set(res))
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/models/product_product.py]: Sorry, just one more change... 
in order to improvement the performance here we should use this transformation just if is really need it.
So IMHO is better use:
```python
if products:
    res = list(set(res + products.name_get()))
return res
```


We can save a lot of seconds by method executed, you could calculate it with:
```bash
python -m timeit -s "res=[(1, 'A'),(2, 'B'),(3, 'C'),(2, 'B'),]" "res=list(set(res))"
1000000 loops, best of 3: 0.596 usec per loop
```

File: product_supplierinfo_for_customer_sale/models/product_product.py [hunk 5]
```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
     def name_search(self, name='', args=None, operator='ilike', limit=80):
         res = super(ProductProduct, self).name_search(
             name, args=args, operator=operator, limit=limit)
-        if res or not self._context.get('partner_id'):
+        if not name or not self._context.get('partner_id'):
             return res
         partner_id = self._context['partner_id']
         supplierinfo = self.env['product.supplierinfo'].search(
             [('name', '=', partner_id), '|', ('product_code', operator, name),
              ('product_name', operator, name)])
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/models/product_product.py]:
```python
if not supplierinfo:
    return res
```

File: product_supplierinfo_for_customer_sale/models/product_product.py (hunk 4 above)
Comment (moylop260) [reply product_supplierinfo_for_customer_sale/models/product_product.py]: Or better...
Use a `product_ids = set(supplierinfo.mapped('product_tmpl_id').ids) - set([id for id, name in res])` from domain of search call method.

And reduce the limit from current res.
`limit=limit - len(res)`

And you will have unique products with exactly the number of records defined and you will save a lot of time reducing the query.

File: product_supplierinfo_for_customer_sale/tests/test_product_name_search.py [hunk 6]
```diff
@@ -58,3 +64,23 @@ def test_10_find_product_customer_code(self):
         self.assertEquals(len(product_names), 1)
         self.assertEquals(self.product.id, product_names[0][0])
         self.assertEquals('[code_product] Name_product', product_names[0][1])
+
+        # Create in product_1 a customer code similar to product, then
+        # name_search must find two products
+        self.assertFalse(self.product_1.customer_ids)
+
+        self.supplierinfo_dict.update({
+            'product_code': 'code_test_1',
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/tests/test_product_name_search.py]: Use the `product.default_code` to return 2 records
One from supplier info and other one from product default search.

**Review by moylop260 — CHANGES_REQUESTED** [review]
sorry other changes

File: product_supplierinfo_for_customer_sale/tests/test_product_name_search.py [hunk 7]
```diff
@@ -19,6 +19,7 @@ def setUp(self):
              'default_code': 'code_product', }).with_context(
                  {'partner_id': self.customer.id,
                   'supplierinfo_type': 'customer'})
+
```
Comment (moylop260) [inline product_supplierinfo_for_customer_sale/tests/test_product_name_search.py]: remove useless change

**Review by moylop260 — APPROVED** [review]
👍 
Thanks!

@JulioSerna 
What do you think about?

**Review by JulioSerna — APPROVED** [review]

**Comment by moylop260** [comment]
@elicoidal 
What do you think about?

**Review by elicoidal — APPROVED** [review]
LGTM

**Pull request merged**